doc/mgmt/SONiC_HAMD_RBAC_HLD.md

```diff
@@ -0,0 +1,1289 @@
+|--------- |----------|------------------ |---------------- | -------------- |
+| 1 | User account with existing user roles | Entry present in USER_DB | Subscription code is not present in HAMD - Hence no entry will be added in the USER_DB | Yes |
+| 2 | User account with new roles added in 4.1 | Entry present in USER_DB | Subscription code is not present in HAMD - Hence no entry will be added in the USER_DB | Yes |
+| | | User account entry in linux kernel | The entry would be persisted by the post script | Yes |
+| | | New groups will be added as a part of the feature | The new groups will be retained and users with the new groups will also be retained. The new roles will not work. The user will behave like an operator. | Yes |
+| 3 | User account with more than one role (support added in 4.1 ) | Entry present in USER_DB | Subscription code is not present in HAMD - Hence no entry will be added in the USER_DB | Yes |
+| | | User account entry in linux kernel | The entry would be persisted by the post script | Yes |
+| 4 | Remote user accounts | Entry present in USER_DB | Radius/Tacacs modules will not write the user details into the USER_DB | Yes |
+| | | New groups will be added as a part of the feature | The new groups will be removed when the remote user logs-in. No functionality issue | Yes |
+| 5 | Pruning the CLI tree | CLI tree will be pruned for both local and remote users | The CLI tree will not be pruned. All the commands will be visible. But when unauthorized commands are executed, the below error will be thrown : "%Error: Client is not authorized to perform this operation"
+
+## 3.13 Resource Needs
+1. New Trie added - memory required  - TODO
+2. Complexity (insertion/deletion/find) will be O(log(n))
+3. The time taken to restart the mgmt-framework and telemetry containers - TODO
+
+# 4 Flow Diagrams
+Refer to section 3. 
+
+# 5 Implementation guide lines
+
+## 5.1 CLI New feature Implementation
+All the commands in XML files under src/sonic-mgmt-framework/CLI/clitree/cli-xml must be updated with the access tag.
+The access field of each XML COMMAND tag must be updated with the appropriate RBAC information.
+For example,
+               <COMMAND name="configure terminal" access="admin:netadmin"
+                        help="Configure from the terminal"
+                        view="configure-view">
+
+## 5.2 RBAC New feature Implementation
+A new extension statement “user-role-priv” is added to specify the RBAC information.
+For all the annotation yangs, the newly defined extension statement should be added along with the other existing extension statements.
+
+For example,
+	deviation /oc-wmr:warm-restart/oc-wmr:config/oc-wmr:enable {
+	    deviate add {
+		sonic-ext:table-name "WARM_RESTART_ENABLE";
+		sonic-ext:key-transformer "wmr_config_en_key_xfmr";
+		sonic-ext:field-name "enable";
+		sonic-ext:user-role-priv "read: admin, netadmin,  secadmin, !operator;  write: admin, netadmin; rpc: admin";
```

Comment:
Why should we add admin role as part of the extension ? By definition, and admin role automatically has all permission to all objects, right ?